### PR TITLE
Add general info metrics package / Add NGT info metrics

### PR DESCRIPTION
**Pull request by rinx**
### Description:

This PR includes two main features.
- Add general info metrics package
    - enables us to add new info metrics easily, based on internal/config structs.
    - added unit tests. already passed with 100% coverage.
- Add NGT info metrics
    - includes "distance type", "object type" and "dimension".
    - using Prometheus exporter, the metrics looks like:
```
# HELP vald_agent_core_ngt_info Agent NGT info
# TYPE vald_agent_core_ngt_info gauge
vald_agent_core_ngt_info{dimension="784",distance_type="l2",object_type="float"} 1
```

### Related Issue:

Nothing

### How Has This Been Tested?:

tested on my k3d cluster.

### Environment:

- Go Version: 1.16
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [X] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [X] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [X] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.
Signed-off-by: Rintaro Okamura <rintaro.okamura@gmail.com>

### CODE REVIEW

File: internal/observability/metrics/info/info.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		default:
+			continue
+		}
+
+		k, err := metrics.NewKey(keyName)
+		if err != nil {
+			return nil, err
+		}
+
+		kvs[k] = value
```
Comment (kpango) [inline internal/observability/metrics/info/info.go:104]: 多分、こんな感じ
```suggestion
                k, err := metrics.NewKey(keyName)
                if err != nil {
                        return nil, err
                }
                switch v := rv.Field(k); v.Kind() {
                case reflect.String:
                        kvs[k] = v.String()
                case reflect.Slice, reflect.Array:
                        kvs[k] = fmt.Sprintf("%.255s", fmt.Sprintf("%v", v.Interface()))
                case reflect.Bool:
                        kvs[k] = strconv.FormatBool(v.Bool())
                case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
                        kvs[k] = strconv.FormatInt(v.Int(), 10)
                case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
                        kvs[k] = strconv.FormatUint(v.Uint(), 10)
                case reflect.Float32, reflect.Float64:
                        kvs[k] = strconv.FormatFloat(v.Float(), 'E', -1, 64)
                }

```

**Review by kpango — CHANGES_REQUESTED** [review]
Others LGTM

File: internal/observability/metrics/info/info.go (hunk 1 above)
Comment (rinx) [reply internal/observability/metrics/info/info.go]: thanks, but if `v` is something other than primitive type or slice of a primitive type (maybe struct or slice of a struct), it should be ignored.

Comment (kpango) [reply internal/observability/metrics/info/info.go]: sorry what do you mean???
for slice value???

Comment (rinx) [reply internal/observability/metrics/info/info.go]: the input value of this function `i` can be any struct type like
```
type A struct {
	B string
	C struct {...}
	D []struct {...}
}
```
.
In this case, C and D should be ignored. `metrics.KeyName` should not be called with these `keyName`s.
In the previous code, `default` branch is for this purpose.

So I think, 
```
                case reflect.Slice, reflect.Array:
```
should be modified to accept only `[]string`.

Comment (rinx) [reply internal/observability/metrics/info/info.go]: Technically it is possible to put complicated struct values into the label. However it is not a good manner for metrics labeling.

Comment (kpango) [reply internal/observability/metrics/info/info.go]: I've fixed the string slice.

However, I haven't set it to kvs for the struct type, but will it affect to the label?

Comment (kpango) [reply internal/observability/metrics/info/info.go]: I noticed that the value of k has been overwritten, you should use the alias of the key used in the for loop

Comment (rinx) [reply internal/observability/metrics/info/info.go]: thanks. I'll revise it.

Just calling `metrics.NewKey(keyName)` makes a new label in the global instance. It is not a good behavior.

Comment (kpango) [reply internal/observability/metrics/info/info.go]: hmm, okay I'll revise it after switch

Comment (rinx) [reply internal/observability/metrics/info/info.go]: there's no need to revise. I'll revise this PR following your code. :)

Comment (kpango) [reply internal/observability/metrics/info/info.go]: I already updated

Comment (rinx) [reply internal/observability/metrics/info/info.go]: revised.

Comment (rinx) [reply internal/observability/metrics/info/info.go]: Is there something wrong? :thinking:

Comment (kpango) [reply internal/observability/metrics/info/info.go]: seems good I'm checking.

Comment (rinx) [reply internal/observability/metrics/info/info.go]: Sorry, I've just added several missing test cases. Please check it :pray:

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**